pkg/templates/nstemplatetiers/nstemplatetier_generator.go

```diff
@@ -0,0 +1,201 @@
+			continue
+		}
+		tierKind := data[0]
+		nsKind := data[1]
+		// create a new entry if needed
+		revision, ok := revision.(string)
+		if !ok {
+			log.Info("invalid namespace template filename revision. Expected a string", "filename", filename, "revision", revision)
+			continue
+		}
+		if _, ok := revisions[tierKind]; !ok {
+			revisions[tierKind] = make(map[string]string, 3) // expect 3 entries: 'code', 'dev' and 'stage'
+		}
+		revisions[tierKind][nsKind] = revision
+	}
+	log.Info("templates revisions loaded", "revisions", revisions)
+	return revisions, nil
+}
+
+// NewNSTemplateTiers generates all manifests, indexed by their associated tier kind and by their namespace kind
+// eg:
+// - advanced:
+//   - code: <[]byte>
+//   - dev: <[]byte>
+//   - stage: <[]byte>
+// - basic:
+//   - code: <[]byte>
+//   - dev: <[]byte>
+//   - stage: <[]byte>
+func (g NSTemplateTierGenerator) NewNSTemplateTiers(namespace string) ([]*toolchainv1alpha1.NSTemplateTier, error) {
+	objects := make([]*toolchainv1alpha1.NSTemplateTier, 0, len(g.revisions))
+	// retrieve the tier names and order them, so we compare them
+	// with the expected templates during the tests
+	tiers := make([]string, 0, len(g.revisions))
+	for tier := range g.revisions {
+		tiers = append(tiers, tier)
+	}
+	sort.Strings(tiers)
+	for _, tier := range tiers {
+		obj, err := g.NewNSTemplateTier(tier, namespace)
+		if err != nil {
+			return nil, errors.Wrapf(err, "unable to generate all NSTemplateTier manifests")
+		}
+		objects = append(objects, obj)
+	}
+	return objects, nil
+}
+
+// NewNSTemplateTier initializes a complete NSTemplateTier object
+// by embedding the `<tier>-code.yml`, `<tier>-dev.yml` and `<tier>-stage.yml`
+// file along with each one's git (short) commit as the revision associated with
+// the template.
+//
+// Something like:
+// ------
+// kind: NSTemplateTier
+//   metadata:
+//     name: basic
+//   spec:
+//     namespaces:
+//     - type: code
+//       revision: y8f907f6
+//       template: >
+//         <yaml-ns-template>
+//     - type: dev
+//       revision: f8q907f4
+//       template: >
+//         <yaml-ns-template>
+//     - type: stage
+//       revision: 907fy8f6
+//       template: >
+//         <yaml-ns-template>
+// ------
+func (g NSTemplateTierGenerator) NewNSTemplateTier(tier, namespace string) (*toolchainv1alpha1.NSTemplateTier, error) {
+	revisions, ok := g.revisions[tier]
+	if !ok {
+		return nil, errors.Errorf("tier '%s' does not exist", tier)
+	}
+	obj := &toolchainv1alpha1.NSTemplateTier{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      tier,
+			Namespace: namespace,
+		},
+		Spec: toolchainv1alpha1.NSTemplateTierSpec{},
+	}
+	// retrieve the namespace types and order them, so we can order them, and compare
+	// with the expected templates during the tests
+	nsTypes := make([]string, 0, len(revisions))
+	for nsType := range revisions {
+		nsTypes = append(nsTypes, nsType)
+	}
+	sort.Strings(nsTypes)
```

Comment:
on the other hand, I understand the purpose of this sorting. In this case you really care about the order in case you want to use `reflect.DeepEqual` function. 
another option would be implementing our own equal function for the `NSTemplateTierSpec` as is done for `NSTemplateSetSpec` https://github.com/codeready-toolchain/api/blob/master/pkg/apis/toolchain/v1alpha1/nstemplateset_types.go#L78
both options are fine, so it's up to you.

Reply:
good point, let me do that instead.

Reply:
I still think it could be useful to have Namespaces sorted in NSTemplateTier.Spec. So we can guarantee that we will get the same yaml (if we marshal the struct) every time we generate a tier for the same templates. It can be useful in prod if we put templates to annotations in the future.